### ISSUE TITLE
Protection against C99 webshell mass exploits

### ISSUE THREAD

**Issue by fzipi** [crstest01/owasp-modsecurity-crs#189]
_Issue originally created by user lifeforms on date 2014-06-24 13:28:39.
            Link to original issue: https://github.com/SpiderLabs/owasp-modsecurity-crs/issues/189._

It turns out that c99.php, a PHP 'web shell', is backdoored by a `c99shcook` parameter that bypasses authentication. http://thehackerblog.com/every-c99-php-shell-is-backdoored-aka-free-shells/

It seems that c99.php itself is also used most often by crackers on exploited web sites, so you might not want to allow files with this name anyway. (I guess it could be simply renamed, but at least that will fend off some mass scans.)

I'd recommend something like the following, which we are using internally:

```
SecRule ARGS_NAMES "c99shcook" \
    "id:6000050,phase:2,t:none,t:lowercase,block,tag:'PHP',msg:'PHP c99 webshell'"

SecRule REQUEST_FILENAME "/c99(\.php)?$" \
    "id:6000051,phase:2,t:none,t:lowercase,block,tag:'PHP',msg:'PHP c99 webshell'"
```

**Comment by fzipi** [comment]
_User quenenni commented on date 2014-07-16 00:49:14:_

I don't think it's the good way to do it.
Some of ours (old) cms on our server were hacked few weeks ago and we found plenty of these webshells, various styles & versions and with plenty of differents names.

We decided to scan all uploaded files with a script (python in our case) a colleague did.

Check this link to see the configurations to do to modsec (in french, but easy to follow):
http://wiki.kogite.fr/index.php/Apache_mod_security,_mod_evasive_:_limitation_du_traffic,_protection_anti-DDOS (the 'mod_security ' paragraph)

And as for the modsec rule:

```
SecRule FILES "(.*)" "phase:2,chain,capture,t:none,t:lowercase,log,auditlog,block,id:90031,logdata:'%{TX.0}',msg:'Custom rule: uploads',setvar:'tx.msg=%{rule.msg}'"
 SecRule FILES_SIZES "@gt 10" "chain"
 SecRule FILES_SIZES "@lt 2000000" "chain"
 SecRule FILES_TMPNAMES "@inspectFile /usr/local/sbin/modsec-scan-upload.py"
```

It takes all uploaded files smaller than 2Mb and scan it with the script.

Smaller than 2mb to avoid scanning audio and video files.
Scan also pictures files as they can be hiding bad stuff.

If you see a way to improve the rule, I'm all ears.

It's also good to know that modsec inspect uploaded files only from multipart/form-data requests.
It means, with my 10 days experience, that it's good for cms like wordpress, drupal, spip, joomla and lots of normal upload forms, but it doesn't work for dokuwiki (I didn't try yet others wiki)

I hope modsec will find a way to be able to inspect all uploaded files in the future.

As for the python script, if you are interested by it, I must first ask my colleague if it's ok for him if I put it here.

If you want to just see all the uploaded files on your server without blocking anything, you can add this rule:

```
SecRule FILES "(.*)" "phase:2,capture,t:none,t:lowercase,log,auditlog,pass,id:90030,logdata:'%{TX.0}',msg:'Custom rule: uploads',setvar:'tx.msg=%{rule.msg}'"
```

**Comment by fzipi** [comment]
_User rcbarnett commented on date 2014-07-16 01:59:40:_

You might want to review this past blog post which shows how to use @inspectFile along with Maldet to detect uploading webshells/backdoors - 
http://blog.spiderlabs.com/2013/10/hiding-webshell-backdoor-code-in-image-files.html

**Comment by fzipi** [comment]
_User quenenni commented on date 2014-07-16 15:13:05:_

I know maldetect. 
But I'm not really convinced by it.

I scanned the folder where we put all the webshells & others bad stuffs we found on our server (158 files) and maldetect found only 14 hits.

That's really too few.

I tested the same dir with our script and it found 154 hits. (3 of the 4 non hit files have just a small piece of php code to send spams, no base64, gzinflate or eval kind of things).
